### PR TITLE
Fix clang static analyzer warning

### PR DESCRIPTION
**Pull request by Nicogene**
For the *strange* `YARP_NORETURN` placement : 
Via Stackoverflow: https://stackoverflow.com/questions/26888805/how-to-declare-a-lambdas-operator-as-noreturn
Clang is correct. An attribute can appertain to a function being declared, or to its type; the two are different. [[noreturn]] must appertain to the function itself. The difference can be seen in
```
// [[noreturn]] appertains to the entity that's being declared
void f [[noreturn]] ();    // §8.3 [dcl.meaning]/p1:
                           // The optional attribute-specifier-seq following a
                           // declarator-id appertains to the entity that is declared."
[[noreturn]] void h ();    // §7 [dcl.dcl]/p2:
                           // "The attribute-specifier-seq in a simple-declaration 
                           // appertains to each of the entities declared by
                           // the declarators of the init-declarator-list."

// ill-formed - [[noreturn]] appertains to the type (§8.3.5 [dcl.fct]/p1: 
// "The optional attribute-specifier-seq appertains to the function type.")
void g () [[noreturn]] {}
```

Indeed if you compile this in g++ it tells you that
```
warning: attribute ignored [-Wattributes]
 void g () [[noreturn]] {}
                      ^
```
note: an attribute that appertains to a type-specifier is ignored
Note that it doesn't emit a warning that g() actually does return.

### CODE REVIEW

File: src/libYARP_OS/include/yarp/os/Log.h [hunk 1]
```diff
@@ -70,7 +70,7 @@ class YARP_OS_API Log
     void info(const char *msg, ...) const YARP_ATTRIBUTE_FORMAT(printf, 2, 3);
     void warning(const char *msg, ...) const YARP_ATTRIBUTE_FORMAT(printf, 2, 3);
     void error(const char *msg, ...) const YARP_ATTRIBUTE_FORMAT(printf, 2, 3);
-    void fatal(const char *msg, ...) const YARP_ATTRIBUTE_FORMAT(printf, 2, 3);
+    void fatal YARP_NORETURN (const char *msg, ...) const YARP_ATTRIBUTE_FORMAT(printf, 2, 3);
```
Comment (traversaro) [inline src/libYARP_OS/include/yarp/os/Log.h]: What is the rational for the unusual placement of this attribute?

Comment (drdanz) [inline src/libYARP_OS/include/yarp/os/Log.h]: Good question, I wasn't even aware that you could add attributes there, but apparently it is the only place where all the compilers like it

Comment (Nicogene) [inline src/libYARP_OS/include/yarp/os/Log.h]: I updated the PR description with the explanation of this placement foundo on Stackoverflow. 
If you don't put it there with clang it doesn't compile.

**Comment by Nicogene** [comment]
This PR will be split in several "little" PR.

**Pull request closed without merging**